chain/consensus/mir/mine.go

```diff
@@ -0,0 +1,185 @@
+package mir
+
+import (
+	"context"
+	"time"
+
+	"github.com/filecoin-project/go-address"
+	lapi "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/api/v1api"
+	"github.com/filecoin-project/lotus/chain/consensus/common"
+	"github.com/filecoin-project/lotus/chain/types"
+	ltypes "github.com/filecoin-project/lotus/chain/types"
+	mirTypes "github.com/filecoin-project/mir/pkg/types"
+)
+
+// Mine mines Filecoin blocks.
+
+// Mine implements the following algorithm:
+// 1. Retrieve messages and cross-messages from mempool.
+// 2. Send these messages to the Mir node.
+// 3. Receive ordered messages from the Mir node and push them into the next Filecoin block.
+// 4. Submit this block.
+func Mine(ctx context.Context, miner address.Address, api v1api.FullNode) error {
+	log.Info("Mir miner started")
+	defer log.Info("Mir miner stopped")
+
+	// TODO: Suppose we want to use Mir in Root and in a subnet at the same time.
+	// Do we need two Mir agents for that?
+	// nodeID, err := api.ID(ctx)
+	// if err != nil {
+	//	log.Fatalf("unable to get a node ID: %s", err)
+	// }
+	nodeID := NodeID()
+	nodeNumber := NodeNumber()
+
+	mirAgent, err := NewMirAgent(nodeID, nodeNumber)
+	if err != nil {
+		return err
+	}
+	mirErrors := mirAgent.Start(ctx)
+	mirHead := mirAgent.App.ChainNotify
+
+	netName, nerr := api.StateNetworkName(ctx)
+	if nerr != nil {
+		return nerr
+	}
+	subnetID := address.SubnetID(netName)
+
+	log.Infof("Mir miner params: network name - %s, subnet ID - %s", netName, subnetID)
+
+	submit := time.NewTicker(SubmitInterval)
```

Comment:
Will this ticker still be needed once Mir is fully implemented? If this is the case, we may want to add a TODO to figure out the optimal ticker to maximize performance.

Reply:
@matejpavlovic will be happy to answer

Reply:
We should not need that ticker later - as soon as the mempool does not need to be polled for new messags. One of the points of the ordering layer architecture that I'm currently drafting is to modify the mempool to be able to create notifications about newly received messages (e.g. by writing them to a channel). When we have that, this timer can go away.

That being said, the underlying protocol ordering the messages will definitely require some form of batching, which, depending on the protocol, might or might not involve some periodic timeouts. But that will be the job of the responsibility of the protocol implementation.